subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -786,99 +819,212 @@ static int load_tls_credential(struct lwm2m_ctx *client_ctx, uint16_t res_id,
+	MBEDTLS_TLS_PSK_WITH_AES_128_CCM_8,
+};
+
+static const int cipher_list_cert[] = {
+	MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+};
+
+#endif /* CONFIG_LWM2M_DTLS_SUPPORT */
+
+int lwm2m_set_default_sockopt(struct lwm2m_ctx *ctx)
+{
 #if defined(CONFIG_LWM2M_DTLS_SUPPORT)
-	else if (client_ctx->use_dtls) {
+	if (ctx->use_dtls) {
+		int ret;
+		uint8_t tmp;
 		sec_tag_t tls_tag_list[] = {
-			client_ctx->tls_tag,
+			ctx->tls_tag,
 		};
 
-		ret = zsock_setsockopt(client_ctx->sock_fd, SOL_TLS, TLS_SEC_TAG_LIST, tls_tag_list,
+		ret = zsock_setsockopt(ctx->sock_fd, SOL_TLS, TLS_SEC_TAG_LIST, tls_tag_list,
 				       sizeof(tls_tag_list));
 		if (ret < 0) {
 			ret = -errno;
 			LOG_ERR("Failed to set TLS_SEC_TAG_LIST option: %d", ret);
-			goto error;
+			return ret;
 		}
 
 		if (IS_ENABLED(CONFIG_LWM2M_TLS_SESSION_CACHING)) {
 			int session_cache = TLS_SESSION_CACHE_ENABLED;
 
-			ret = zsock_setsockopt(client_ctx->sock_fd, SOL_TLS, TLS_SESSION_CACHE,
+			ret = zsock_setsockopt(ctx->sock_fd, SOL_TLS, TLS_SESSION_CACHE,
 					       &session_cache, sizeof(session_cache));
 			if (ret < 0) {
 				ret = -errno;
 				LOG_ERR("Failed to set TLS_SESSION_CACHE option: %d", errno);
-				goto error;
+				return ret;
 			}
 		}
 
-		if (client_ctx->hostname_verify && (client_ctx->desthostname != NULL)) {
+		if (ctx->hostname_verify && (ctx->desthostname != NULL)) {
 			/** store character at len position */
-			tmp = client_ctx->desthostname[client_ctx->desthostnamelen];
+			tmp = ctx->desthostname[ctx->desthostnamelen];
 
 			/** change it to '\0' to pass to socket*/
-			client_ctx->desthostname[client_ctx->desthostnamelen] = '\0';
+			ctx->desthostname[ctx->desthostnamelen] = '\0';
 
 			/** mbedtls ignores length */
-			ret = zsock_setsockopt(client_ctx->sock_fd, SOL_TLS, TLS_HOSTNAME,
-					       client_ctx->desthostname,
-					       client_ctx->desthostnamelen);
+			ret = zsock_setsockopt(ctx->sock_fd, SOL_TLS, TLS_HOSTNAME,
+					       ctx->desthostname, ctx->desthostnamelen);
 
 			/** restore character */
-			client_ctx->desthostname[client_ctx->desthostnamelen] = tmp;
+			ctx->desthostname[ctx->desthostnamelen] = tmp;
 			if (ret < 0) {
 				ret = -errno;
 				LOG_ERR("Failed to set TLS_HOSTNAME option: %d", ret);
-				goto error;
+				return ret;
+			}
+
+			int verify = TLS_PEER_VERIFY_REQUIRED;
+
+			ret = zsock_setsockopt(ctx->sock_fd, SOL_TLS, TLS_PEER_VERIFY, &verify,
+					       sizeof(verify));
+			if (ret) {
+				LOG_ERR("Failed to set TLS_PEER_VERIFY");
+			}
+
+		} else {
+			/* By default, Mbed TLS tries to verify peer hostname, disable it */
+			int verify = TLS_PEER_VERIFY_NONE;
+
+			ret = zsock_setsockopt(ctx->sock_fd, SOL_TLS, TLS_PEER_VERIFY, &verify,
+					       sizeof(verify));
+			if (ret) {
+				LOG_ERR("Failed to set TLS_PEER_VERIFY");
 			}
 		}
+
+		switch (lwm2m_security_mode(ctx)) {
+		case LWM2M_SECURITY_PSK:
+			ret = zsock_setsockopt(ctx->sock_fd, SOL_TLS, TLS_CIPHERSUITE_LIST,
```

Comment:
Why do we need to narrow the list of cipher suites to use, instead of letting mbedTLS to choose?

Reply:
Because when all PSK and certificate compatible cipher suites were enabled, MbedTLS tried to offer all to Client HELLO causing the handshake to always start with certificate mode.
I had to limit that in runtime to force it to PSK mode.

Reply:
Interesting, I thought mbed TLS would be smart enough not to offer certificate-based suites if there's no cerfiticate registered for the context. Ok then.

Reply:
I can see that there is difference on how MbedTLS works on Zephyr, and how nRF9160 modem works.
Looks like TLS storage content is not checked before the negotiation starts, or at least that is what I assume.
But on Nordic modem, what ciphers are offered, seem to depend on the content of the security tag.

Therefore only portable choice seem to be to select the offered cipher-suites by hand.